google/cloud/bigquery/external_config.py

```diff
@@ -1003,3 +1003,104 @@ def from_api_repr(cls, resource: dict) -> "ExternalConfig":
         config = cls(resource["sourceFormat"])
         config._properties = copy.deepcopy(resource)
         return config
+
+
+class ExternalCatalogDatasetOptions(object):
+    """Options defining open source compatible datasets living in the BigQuery catalog.
+    Contains metadata of open source database, schema or namespace represented
+    by the current dataset.
+
+    Args:
+        default_storage_location_uri: Optional. The storage location URI for all
+            tables in the dataset. Equivalent to hive metastore's database
+            locationUri. Maximum length of 1024 characters. (str)
+        parameters: Optional. A map of key value pairs defining the parameters
+            and properties of the open source schema. Maximum size of 2Mib.
+    """
+
+    def __init__(
+        self,
+        default_storage_location_uri: Optional[str] = None,
```

Comment:
I think it would be good if we add properties for these values too. I wonder if we should also provide a property setter, or make it read only.

Reply:
@Linchin 

In thinking about this...

I think we want/need a different approach.

Rather than having all the input validation in the `__init__()` method, I think I wanna have `__init__()` call the setter and let the setter do the input validation so that I only need that code in one place.

Thoughts?

This `Table` class in the `table.py` file does this with the schema argument:

```
if schema is not None:
    self.schema = schema
```

and relies on input validation here:

```
@schema.setter
def schema(self, value):
    api_field = self._PROPERTY_TO_API_FIELD["schema"]

    if value is None:
        self._properties[api_field] = None
    else:
        value = _to_schema_fields(value)
        self._properties[api_field] = {"fields": _build_schema_resource(value)}
```

https://github.com/googleapis/python-bigquery/blob/255472359f3ed6b6cee06039ebe9059607fd9894/google/cloud/bigquery/table.py#L355C1-L417C1

Reply:
I think this is a great idea! This way we can reduce redundancy significantly.